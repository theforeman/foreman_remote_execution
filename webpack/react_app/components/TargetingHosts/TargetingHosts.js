import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import { LoadingState, Alert } from 'patternfly-react';
import { STATUS } from 'foremanReact/constants';
import HostItem from './components/HostItem';

const TargetingHosts = ({ status, items }) => {
  if (status === STATUS.ERROR) {
    return (
      <Alert type="error">
        {__(
          'There was an error while updating the status, try refreshing the page.'
        )}
      </Alert>
    );
  }

  return (
    <LoadingState loading={!items.length}>
      <div>
        <table className="table table-bordered table-striped table-hover">
          <thead>
            <tr>
              <th>{__('Host')}</th>
              <th>{__('Status')}</th>
              <th>{__('Actions')}</th>
            </tr>
          </thead>
          <tbody>
            {items.map(host => (
              <HostItem
                key={host.name}
                name={host.name}
                link={host.link}
                status={host.status}
                actions={host.actions}
              />
            ))}
          </tbody>
        </table>
      </div>
    </LoadingState>
  );
};

TargetingHosts.propTypes = {
  status: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
};

export default TargetingHosts;
