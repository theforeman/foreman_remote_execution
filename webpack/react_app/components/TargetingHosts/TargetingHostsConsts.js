export const TARGETING_HOSTS = 'TARGETING_HOSTS';
