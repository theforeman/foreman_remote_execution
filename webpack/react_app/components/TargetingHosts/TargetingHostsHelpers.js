import { getURI } from 'foremanReact/common/urlHelpers';

export const getApiUrl = (searchQuery, pagination) => {
  const baseUrl = getURI()
    .search('')
    .addQuery('page', pagination.page)
    .addQuery('per_page', pagination.perPage);

  return searchQuery ? baseUrl.addQuery('search', searchQuery) : baseUrl;
};
