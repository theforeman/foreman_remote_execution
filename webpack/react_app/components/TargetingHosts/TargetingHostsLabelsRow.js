import React from 'react';
import PropTypes from 'prop-types';
import { Row, Label } from 'patternfly-react';
import { noop } from 'foremanReact/common/helpers';
import { translate as __ } from 'foremanReact/common/I18n';

import './TargetingHostsLabelsRow.scss';

const TargetingHostsLabelsRow = ({ query, updateQuery }) => {
  const onDeleteClick = keyToDelete => {
    const { [keyToDelete]: deleted, ...queryWithoutDeleted } = query;
    updateQuery(queryWithoutDeleted);
  };

  const queryEntries = Object.entries(query);

  return (
    queryEntries.length > 0 && (
      <Row className="tasks-labels-row">
        <span className="title">{__('Active Filters:')}</span>
        {queryEntries.map(([key, value]) => (
          <Label
            bsStyle="info"
            key={key}
            onRemoveClick={() => onDeleteClick(key)}
          >
            {`${key} = ${value}`}
          </Label>
        ))}
      </Row>
    )
  );
};

TargetingHostsLabelsRow.propTypes = {
  query: PropTypes.object,
  updateQuery: PropTypes.func,
};

TargetingHostsLabelsRow.defaultProps = {
  query: {},
  updateQuery: noop,
};

export default TargetingHostsLabelsRow;
