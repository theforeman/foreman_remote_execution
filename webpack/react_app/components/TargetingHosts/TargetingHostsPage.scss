#targeting_hosts {
  min-height: 350px;
}
