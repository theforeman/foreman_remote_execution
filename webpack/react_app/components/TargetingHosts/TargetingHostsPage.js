import React from 'react';
import PropTypes from 'prop-types';
import { Grid } from 'patternfly-react';

import SearchBar from 'foremanReact/components/SearchBar';
import Pagination from 'foremanReact/components/Pagination/PaginationWrapper';
import { getControllerSearchProps } from 'foremanReact/constants';

import TargetingHosts from './TargetingHosts';
import './TargetingHostsPage.scss';

const TargetingHostsPage = ({
  handleSearch,
  searchQuery,
  apiStatus,
  items,
  totalHosts,
  pagination,
  handlePagination,
}) => (
  <div id="targeting_hosts">
    <Grid.Row>
      <Grid.Col md={6} className="title_filter">
        <SearchBar
          onSearch={query => handleSearch(query)}
          data={{
            ...getControllerSearchProps('hosts'),
            autocomplete: {
              id: 'targeting_hosts_search',
              searchQuery,
              url: '/hosts/auto_complete_search',
              useKeyShortcuts: true,
            },
            bookmarks: {},
          }}
        />
      </Grid.Col>
    </Grid.Row>
    <br />
    <TargetingHosts apiStatus={apiStatus} items={items} />
    <Pagination
      viewType="table"
      itemCount={totalHosts}
      pagination={pagination}
      onChange={args => handlePagination(args)}
      dropdownButtonId="targeting-hosts-pagination-dropdown"
      className="targeting-hosts-pagination"
    />
  </div>
);

TargetingHostsPage.propTypes = {
  handleSearch: PropTypes.func.isRequired,
  searchQuery: PropTypes.string.isRequired,
  apiStatus: PropTypes.string,
  items: PropTypes.array.isRequired,
  totalHosts: PropTypes.number.isRequired,
  pagination: PropTypes.object.isRequired,
  handlePagination: PropTypes.func.isRequired,
};

TargetingHostsPage.defaultProps = {
  apiStatus: null,
};

export default TargetingHostsPage;
