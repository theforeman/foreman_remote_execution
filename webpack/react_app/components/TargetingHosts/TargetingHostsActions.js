import { getURI } from 'foremanReact/common/urlHelpers';
import { get } from 'foremanReact/redux/API';
import { withInterval } from 'foremanReact/redux/middlewares/IntervalMiddleware';
import { TARGETING_HOSTS } from './TargetingHostsConsts';

const url = getURI().addQuery('format', 'json');
export const getData = () =>
  withInterval(get({ key: TARGETING_HOSTS, url }), 1000);
