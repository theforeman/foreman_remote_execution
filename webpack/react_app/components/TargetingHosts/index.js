import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { get } from 'foremanReact/redux/API';
import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
import {
  withInterval,
  stopInterval,
} from 'foremanReact/redux/middlewares/IntervalMiddleware';

import {
  selectItems,
  selectApiStatus,
  selectAutoRefresh,
  selectTotalHosts,
  selectIntervalExists,
} from './TargetingHostsSelectors';
import { getApiUrl } from './TargetingHostsHelpers';
import { TARGETING_HOSTS } from './TargetingHostsConsts';
import TargetingHostsPage from './TargetingHostsPage';

const WrappedTargetingHosts = () => {
  const dispatch = useDispatch();
  const { perPage, perPageOptions } = useForemanSettings();

  const autoRefresh = useSelector(selectAutoRefresh);
  const items = useSelector(selectItems);
  const apiStatus = useSelector(selectApiStatus);
  const totalHosts = useSelector(selectTotalHosts);
  const [searchQuery, setSearchQuery] = useState('');
  const [pagination, setPagination] = useState({
    page: 1,
    perPage,
    perPageOptions,
  });
  const [apiUrl, setApiUrl] = useState(getApiUrl(searchQuery, pagination));
  const intervalExists = useSelector(selectIntervalExists);

  const handleSearch = query => {
    const defaultPagination = { page: 1, perPage: pagination.perPage };
    stopApiInterval();

    setApiUrl(getApiUrl(query, defaultPagination));
    setSearchQuery(query);
    setPagination(defaultPagination);
  };

  const handlePagination = args => {
    stopApiInterval();
    setPagination(args);
    setApiUrl(getApiUrl(searchQuery, args));
  };

  const stopApiInterval = () => {
    if (intervalExists) {
      dispatch(stopInterval(TARGETING_HOSTS));
    }
  };

  const getData = url =>
    withInterval(
      get({
        key: TARGETING_HOSTS,
        url,
        handleError: () => {
          dispatch(stopInterval(TARGETING_HOSTS));
        },
      }),
      1000
    );

  useEffect(() => {
    dispatch(getData(apiUrl));

    if (autoRefresh === 'false') {
      dispatch(stopInterval(TARGETING_HOSTS));
    }

    return () => {
      dispatch(stopInterval(TARGETING_HOSTS));
    };
  }, [dispatch, apiUrl, autoRefresh]);

  return (
    <TargetingHostsPage
      handleSearch={handleSearch}
      searchQuery={searchQuery}
      apiStatus={apiStatus}
      items={items}
      totalHosts={totalHosts}
      pagination={pagination}
      handlePagination={handlePagination}
    />
  );
};

export default WrappedTargetingHosts;
