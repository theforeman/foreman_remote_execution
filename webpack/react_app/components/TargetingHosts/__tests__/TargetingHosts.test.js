import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import TargetingHosts from '../TargetingHosts';
import { TargetingHostsFixtures } from './fixtures';

describe('TargetingHosts', () =>
  testComponentSnapshotsWithFixtures(TargetingHosts, TargetingHostsFixtures));
