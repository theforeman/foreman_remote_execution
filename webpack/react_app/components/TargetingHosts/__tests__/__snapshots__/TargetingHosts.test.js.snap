// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TargetingHosts renders 1`] = `
<LoadingState
  additionalClasses=""
  loading={false}
  loadingText="Loading"
  size="lg"
  timeout={300}
>
  <div>
    <table
      className="table table-bordered table-striped table-hover"
    >
      <thead>
        <tr>
          <th>
            Host
          </th>
          <th>
            Status
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <HostItem
          actions={Array []}
          key="host"
          link="/link"
          name="host"
          status="success"
        />
      </tbody>
    </table>
  </div>
</LoadingState>
`;

exports[`TargetingHosts renders with error 1`] = `
<Alert
  className=""
  onDismiss={null}
  type="error"
>
  There was an error while updating the status, try refreshing the page.
</Alert>
`;

exports[`TargetingHosts renders with loading 1`] = `
<LoadingState
  additionalClasses=""
  loading={true}
  loadingText="Loading"
  size="lg"
  timeout={300}
>
  <div>
    <table
      className="table table-bordered table-striped table-hover"
    >
      <thead>
        <tr>
          <th>
            Host
          </th>
          <th>
            Status
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</LoadingState>
`;
